### ISSUE TITLE
[Reproducibility] Released MGLL-Fundus data do not match the loader; smooth-KL can be silently omitted

### ISSUE THREAD

**Issue by nthng-quan** [HUANGLIZI/MGLL#3]
## Summary

We audited MGLL-Fundus pretraining at commit [`f7cae639843d2de129b8a7bb0fd7dd36ee109421`](https://github.com/HUANGLIZI/MGLL/tree/f7cae639843d2de129b8a7bb0fd7dd36ee109421) and attempted to train `MultiGranularModule` from the released raw images and label JSON files.

Four repository/data issues prevent a verbatim MGLL-Fundus run:

1. `models/__init__.py` imports two source files absent from the commit.
2. `PretrainMMFundusDataset` expects NPZ files and list files that are not part of the released data.
3. `kl_loss_2` takes `.log()` of raw similarity logits. When this produces a non-finite value, the forward method silently excludes the consistency term.

Separately, on the H100 software stack below, the upstream FP16/autocast/ `GradScaler` path skipped all 20 attempted optimizer updates in a controlled test. This fifth finding is environment-specific; the first four are directly visible in the pinned source and released-data contract.

After adding a manifest adapter, computing KL on softmax probabilities, failing on non-finite losses/gradients, and using FP32, we completed 20 epochs with finite weights and populated AdamW state. Because this changes the input adapter, objective implementation, and precision path, it should be described as a **corrected run of the released model implementation**, not a bit-for-bit reproduction of the reported pretrained weights.

## Data provenance and scope

`mgll_released_split_manifest.json` is **our locally generated manifest** built from the authors' released image and label folders. It is not an artifact present in the pinned GitHub repository. The manifest adapter and the counts below therefore validate our conversion of the released raw data; they do not show that the authors trained from this manifest.

The local manifest contains:

```text
split       image records
train       225,326
validation   14,132
test          6,929
total       246,387
```

Only train and validation annotations were used for pretraining.

## Environment

```text
OS: Linux 5.15.0-1078-nvidia, glibc 2.39
Python: 3.11.15
PyTorch: 2.5.0+cu124
torchvision: 0.20.0+cu124
Transformers: 4.46.3
open-clip-torch: 2.32.0
sentencepiece: 0.2.2
CUDA runtime: 12.4
cuDNN: 9.1.0
GPU: NVIDIA H100 80GB HBM3 MIG 7g.80gb
GPU compute capability: 9.0
Environment manager: uv
Repository commit: f7cae639843d2de129b8a7bb0fd7dd36ee109421
```

## Expected behavior

An MGLL-Fundus pretraining run using `MultiGranularModule` should:

- load the released train and validation data after documented preprocessing;
- optimize image/disease and image/description alignment;
- optimize the cross-granularity consistency term;
- execute each scheduled AdamW update; and
- save checkpoints with non-empty AdamW state and advancing step counters.

### 1. Missing source files prevent package import

At the pinned commit, [`models/__init__.py`](https://github.com/HUANGLIZI/MGLL/blob/f7cae639843d2de129b8a7bb0fd7dd36ee109421/models/__init__.py)
contains:

```python
from .llama import ModelArgs, Transformer
from .utils import format_prompt
```

Neither `models/llama.py` nor `models/utils.py` exists in that commit. Python executes `models/__init__.py` before resolving `models.pretrain`, so even a BERT-only import fails.

For our BERT run, we removed only these two unavailable imports. This does not restore the LLaMA path; its implementation and checkpoint contract remain unavailable.

### 2. The released raw data do not satisfy the implemented loader contract

[`PretrainMMFundusDataset`](https://github.com/HUANGLIZI/MGLL/blob/f7cae639843d2de129b8a7bb0fd7dd36ee109421/util/datasets.py)
expects:

- `<json_list>_train.txt` and `<json_list>_val.txt`;
- each line in those files to name an annotation JSON file;
- images preconverted into `All_data_npz/`; and
- image paths reconstructed with `url[9:]`.

The released folders instead contain raw images and per-dataset label JSON files. The repository does not provide the preprocessing program or list files needed to construct the loader's expected layout.

The loader also creates a LLaMA SentencePiece tokenizer even when the selected model uses BERT, requiring a tokenizer file that is irrelevant to this path.

We added `ManifestFundusDataset`, which reads our locally built manifest, resolves raw RGB images, flattens multiple annotations per image, and returns the same three-item batch contract:

```python
(image, disease_text, description_text)
```

### 3. Smooth-KL is evaluated on raw logits and silently dropped

The released two-granularity implementation in [`models/pretrain.py`](https://github.com/HUANGLIZI/MGLL/blob/f7cae639843d2de129b8a7bb0fd7dd36ee109421/models/pretrain.py) is:

```python
def kl_loss_2(self, pred_logit_1, pred_logit_2):
    M = 0.5 * (pred_logit_1 + pred_logit_2)
    return (
        0.5 * F.kl_div(pred_logit_1.log(), M, reduction="batchmean")
        + 0.5 * F.kl_div(pred_logit_2.log(), M, reduction="batchmean")
    )
```

Similarity logits are neither non-negative nor normalized distributions. Consequently, `.log()` is undefined for negative logits, and `M` is not a probability target.

The forward method then hides the failure:

```python
distill_loss = self.kl_loss_2(logits_per_text, logits_per_desc) * 0.5
if not torch.isnan(distill_loss) and not torch.isinf(distill_loss):
    clip_loss += distill_loss
```

Training can therefore print a finite total loss while omitting the defining cross-granularity term.

The paper defines predicted distributions `P_i`, their mean `M`, and `L_sKL = sum_i KL(P_i || M)`. Our numerical correction was:

```python
log_p1 = F.log_softmax(logits_per_text.float(), dim=-1)
log_p2 = F.log_softmax(logits_per_desc.float(), dim=-1)
p1, p2 = log_p1.exp(), log_p2.exp()
log_mean = (0.5 * (p1 + p2)).clamp_min(1e-8).log()

smooth_kl = (
    F.kl_div(log_mean, p1, reduction="batchmean")
    + F.kl_div(log_mean, p2, reduction="batchmean")
)
```

With PyTorch's argument convention, these calls compute `KL(p1 || M)` and `KL(p2 || M)`. We also replaced the silent skip with a finite-value assertion.

### Additional configuration discrepancy

`main_pretrain.py` exposes `--clip_grad`, but `engine_pretrain.py` ignores its value and passes `clip_grad=10` directly to the scaler on every update. This did not block our run, but it makes the effective configuration differ from the CLI arguments and should be clarified or wired to `args.clip_grad`.

## Environment-specific AMP failure

After applying only the import fix and released-data adapter, we ran the upstream objective and precision path for 20 batches:

```text
module: MultiGranularModule
visual encoder: clip_vit
text encoder: bert
batch size: 4
training samples: 80
training batches: 20
optimizer: AdamW
learning rate: 1e-4
```

The console output appeared normal:

```text
[ 0/20] cliploss: 4.8202
[19/20] cliploss: 5.5274
Training time 0:00:18
```

The saved checkpoint instead contained:

```text
optimizer state entries: 0
GradScaler scale: 0.0625
model tensors: 647
non-finite model tensors: 0
```

The default scale was 65,536, and `65536 * 0.5^20 = 0.0625`. The scaler thus backed off after every attempted update. `GradScaler.step()` skips `optimizer.step()` when it detects non-finite gradients, and AdamW creates its state lazily on the first real step. Empty AdamW state therefore demonstrates that no optimizer update occurred in this 20-batch test.

Minimal checkpoint audit:

```python
import torch

checkpoint = torch.load("checkpoint-0.pth", map_location="cpu")
print(len(checkpoint["optimizer"]["state"]))
print(checkpoint["scaler"])
```

Observed:

```python
0
{
    "scale": 0.0625,
    "growth_factor": 2.0,
    "backoff_factor": 0.5,
    "growth_interval": 2000,
    "_growth_tracker": 0,
}
```

This proves failure on the environment listed above. It does not establish that the AMP path fails on every GPU or software version.

### Compatibility and data

1. Removed the unavailable `models.llama` and `models.utils` imports for the BERT path.
2. Added `--manifest`, `--project_root`, `--train_limit`, and `--val_limit`.
3. Added a raw-image manifest adapter with the existing three-item batch contract.
4. Created a locked `uv` environment. At the pinned commit, `requirements.txt` pins PyTorch but leaves most other versions unpinned; it lists `sentencepiece` (not `sentencepiece7`).

### Objective and numerical behavior

1. Replaced raw-logit `.log()` KL with probability-space smooth-KL.
2. Replaced the silent non-finite consistency-loss skip with an exception.
3. Returned and logged the pre-existing disease, description, and consistency contributions separately.
4. Disabled autocast, explicit image `.half()`, and GradScaler; training ran in FP32.
5. Added a finite-gradient check before each optimizer step.

### Run configuration

```text
module: MultiGranularModule
visual encoder: clip_vit (ViT-L/14)
text encoder: bert (BioClinicalBERT)
batch size: 32
epochs: 20
warmup epochs: 1
optimizer: AdamW
peak learning rate: 1e-4
weight decay: 1e-4
schedule: half-cycle cosine decay
seed: 0
```

Relative to the values in `exps/pretrain.sh`, we retained batch size, epoch count, warmup, optimizer, learning rate, weight decay, augmentations, and checkpoint schedule. The module, text encoder, data path, KL computation, and precision path changed as described above.

Later downstream-evaluation compatibility changes in our worktree were not used by pretraining and are outside this report.

## Validation after correction

Epoch 0 was completed as an acceptance run and epochs 1-19 were resumed from
that checkpoint, for 20 total epochs:

```text
final optimizer step: 145,420
optimizer state entries: 495
states with nonzero momentum: 495/495
non-finite model tensors: 0
epoch-1 train loss: 2.8367
epoch-20 train loss: 2.3294
best validation clip score: 0.8276 at epoch 16 (one-indexed)
```

The reported `clip score` is the repository's internal `contrastive_accuracy` validation metric. It is not a downstream AUC result and is not presented as reproduction of a paper benchmark.

## Data caveats

Our generated manifest contains 246,387 image records, two fewer than the paper's reported 246,389 images. Because some records contain multiple annotations, flattened annotation counts are larger and should not be compared directly with the paper's image count.

An audit of our manifest also found:

```text
REFUGE annotations containing "NormalAbnormal": 71
empty descriptions: 72
```

We preserved these annotations in the reported run. These issues are separate from the import, loader, KL, and optimizer-update failures, but they can prevent exact data-level reproduction.

## Questions

1. Can you provide `models/llama.py` and `models/utils.py`, or make those imports optional for non-LLaMA runs?
2. Can you provide the preprocessing script, exact train/validation split lists, and path convention used to create `All_data_npz/` from the released raw data?
3. What exact PyTorch, CUDA, OpenCLIP, Transformers, and GPU versions were used for reported pretraining?
4. Should `kl_loss_2` operate on softmax probabilities as specified by Eqs. 5-6?
5. Should the effective smooth-KL coefficient be `1.0` as specified by Eq. 7, or `0.5` as applied by the GitHub forward method?
6. Is the hard-coded gradient clipping value of `10` intentional, and should `engine_pretrain.py` use `args.clip_grad`?
